thrimshim/thrimshim/main.py

```diff
@@ -0,0 +1,186 @@
+		return 'Row id = {} not found'.format(ident), 404
+	assert row.id == ident
+	response = row._asdict()
+
+	response = {
+		key: (
+			value.isoformat() if isinstance(value, datetime.datetime)
+			else value
+		) for key, value in response.items()
+	}
+	return json.dumps(response)
+
+
+
+def update_row(ident, new_row):
+	"""Updates row of database with id = ident with the edit columns in
+	new_row.
+
+	If a 'video_link' is provided in update, interperet this as a manual video
+	upload and set state to 'DONE'"""
+
+	edit_columns = ['allow_holes', 'uploader_whitelist', 'upload_location',
+				'video_start', 'video_end', 'video_title', 'video_description',
+				'video_channel', 'video_quality']
+	state_columns = ['state', 'uploader', 'error', 'video_link'] 
+	#these have to be set before a video can be set as 'EDITED'
+	non_null_columns = ['upload_location', 'video_start', 'video_end',
+		'video_channel', 'video_quality']
+
+	#check vital edit columns are in new_row
+	wanted = set(non_null_columns + ['state'])
+	missing = wanted - set(new_row)
+	if missing:
+		return 'Fields missing in JSON: {}'.format(', '.join(missing)), 400
+	#get rid of irrelevant columns
+	extras = set(new_row) - set(edit_columns + state_columns)
+	for extra in extras:
+		del new_row[extra]
+
+	conn = app.db_manager.get_conn()
+	#check a row with id = ident is in the database
+	with database.transaction(conn):
+		results = database.query(conn, """
+			SELECT id, state 
+			FROM events
+			WHERE id = %s;""", ident)
+	old_row = results.fetchone()
+	if old_row is None:
+		return 'Row {} not found'.format(ident), 404
+	assert old_row.id == ident
+	
+	if old_row.state not in ['UNEDITED', 'EDITED', 'CLAIMED']:
+		return 'Video already published', 403
+	
+	# handle state columns
+	# handle non-empty video_link as manual uploads
+	# otherwise clear other state columns
+	if 'video_link' in new_row and new_row['video_link']:
+		new_row['state'] = 'DONE'
+		new_row['upload_location'] = 'manual'
+	else:
+		if new_row['state'] == 'EDITED':
+			missing = []
+			for column in non_null_columns:
+				if not new_row[column] or new_row[column] is None:
+					missing.append(column)
+			if missing:
+				return 'Fields {} must be non-null for video to be cut'.format(', '.join(missing)), 400
+		elif new_row['state'] != 'UNEDITED':
+			return 'Invalid state {}'.format(new_row['state']), 400
+		
+	new_row['uploader'] = None
+	new_row['error'] = None
+
+	# actually update database
+	build_query = sql.SQL("""
+		UPDATE events
+		SET {}
+		WHERE id = %(id)s
+		AND state IN ('UNEDITED', 'EDITED', 'CLAIMED')"""
+		).format(sql.SQL(", ").join(
+			sql.SQL("{} = {}").format(
+				sql.Identifier(column), sql.Placeholder(column),
+			) for column in new_row.keys()
+		))
+	with database.transaction(conn):
+		result = database.query(conn, build_query, id=ident, **new_row)
+	if result.rowcount != 1:
+		if result.rowcount == 0.:
+			with database.transaction(conn):
+				check_result = database.query(conn, """
```

Comment:
I think this second query's completely unnessecary, and we can assume that if our update fails, it's because the video is published.